File: mobile/lib/tradeScorecard.js
// Founder-requested 2026-08-20: a small Executive Briefing card showing how many trades
// each day, week and month were successful and how many were not, with a short AI summary
// of the lessons learned.
//
// The reshaping rules here exist because of specific things that have gone wrong on this
// screen before:
//   - `null` (not yet loaded) must never render as "0 successful". Number(null) === 0 is
//     finite, and that exact trap already produced a card asserting "Low confidence" when
//     the truth was "unknown" (see marketForecast.js's confidenceLabel).
//   - Trades that closed but have no reconciled profit or loss are shown separately as
//     "awaiting reconciliation", never folded into the win or loss counts.
//   - The lessons line is kept short. The briefing already suffers from long generated
//     text burying the short high-value sections.

// 2026-08-25 Founder-reported: the briefing said "Today: 0 worked / 1 didn't" while the
// Portfolio card on the same refresh said "Completed Trades Today: 0". Neither number was
// wrong -- they answer different questions while wearing the same word. These buckets are
// ROLLING windows (see summarize_trade_outcomes: last 24h / 7d / 30d, deliberately rolling so
// the card is not near-empty just after a calendar month turns over), whereas the Portfolio
// card counts the calendar day since midnight. At 2pm the rolling window still contains
// yesterday afternoon, so the two legitimately disagree.
//
// Labelling them honestly is the fix, not forcing them to match: a founder who reads "Last 24
// hours" and "Completed today" can hold both in his head at once, where "Today" and "Today"
// showing different numbers just means the app cannot be trusted.
const PERIOD_LABELS = [
  ['day', 'Last 24 hours'],
  ['week', 'Last 7 days'],
  ['month', 'Last 30 days'],
];

const NO_SCORECARD_ROWS = [];

const NO_SCORECARD_MESSAGE = 'Trade results have not loaded yet.';

function isFiniteNumber(value) {
  if (value === null || value === undefined || value === '') return false;
  return Number.isFinite(Number(value));
}

function countText(bucket) {
  if (!bucket) return null;
  const successful = isFiniteNumber(bucket.successful) ? Number(bucket.successful) : null;
  const unsuccessful = isFiniteNumber(bucket.unsuccessful) ? Number(bucket.unsuccessful) : null;
  if (successful === null || unsuccessful === null) return null;
  if (successful === 0 && unsuccessful === 0) return 'No completed trades';
  return `${successful} worked / ${unsuccessful} didn't`;
}

function winRateText(bucket) {
  if (!bucket || !isFiniteNumber(bucket.win_rate)) return null;
  return `${Math.round(Number(bucket.win_rate) * 100)}% win rate`;
}

function pendingText(bucket) {
  if (!bucket || !isFiniteNumber(bucket.unknown)) return null;
  const unknown = Number(bucket.unknown);
  if (unknown <= 0) return null;
  return `${unknown} awaiting reconciliation`;
}

function netText(bucket) {
  if (!bucket || !isFiniteNumber(bucket.net_pnl)) return null;
  const net = Number(bucket.net_pnl);
  if (net === 0) return null;
  const sign = net > 0 ? '+' : '-';
  return `${sign}£${Math.abs(net).toFixed(2)}`;
}

function scorecardRows(scorecard) {
  if (!scorecard) return NO_SCORECARD_ROWS;
  return PERIOD_LABELS.map(([key, label]) => {
    const bucket = scorecard[key];
    return {
      key,
      label,
      counts: countText(bucket) || 'No completed trades',
      winRate: winRateText(bucket),
      pending: pendingText(bucket),
      net: netText(bucket),
    };
  });
}

function feesText(scorecard) {
  const fees = scorecard && scorecard.fees;
  if (!fees || !fees.available) return null;
  return typeof fees.plain_english === 'string' && fees.plain_english.trim()
    ? fees.plain_english.trim()
    : null;
}

function lessonsText(scorecard) {
  if (!scorecard) return NO_SCORECARD_MESSAGE;
  const lessons = typeof scorecard.lessons === 'string' ? scorecard.lessons.trim() : '';
  if (!lessons) return 'No lessons recorded yet.';
  return lessons;
}

function tradeScorecardCard(scorecard) {
  return {
    loaded: Boolean(scorecard),
    rows: scorecardRows(scorecard),
    lessons: lessonsText(scorecard),
    fees: feesText(scorecard),
  };
}

module.exports = {
  NO_SCORECARD_MESSAGE,
  NO_SCORECARD_ROWS,
  countText,
  feesText,
  lessonsText,
  netText,
  pendingText,
  scorecardRows,
  tradeScorecardCard,
  winRateText,
};

File: src/ai_trader/trade_scorecard.py
"""Founder-facing trade scorecard: how many trades worked, how many didn't, and what
was actually learned -- Founder-requested 2026-08-20.

*"I would like to see a small card on the executive briefing screen with how many trades
each day, week and month were successful and how many were not with them a short ai
summary of one or two sentences on the lessons learned."*

Two honesty rules are baked in rather than left to the caller:

1. A closed trade with no reconciled P&L is counted as `unknown`, never silently folded
   into wins or losses. The Founder has been shown confident-looking numbers built on
   absent data before; a scorecard that quietly rounds unknowns into "successful" is worse
   than no scorecard.
2. The lessons line is short by construction (the Founder asked for one or two sentences),
   and when there is genuinely nothing to learn from it says so plainly instead of
   generating filler. The existing Executive Briefing already suffers from verbose
   generated text crowding out the short high-value sections.
"""

from __future__ import annotations

import sqlite3
from contextlib import closing
from datetime import datetime, timezone
from pathlib import Path
from typing import Any, Iterable

from .database import connect


_PERIODS: dict[str, float] = {
    "day": 86_400.0,
    "week": 7 * 86_400.0,
    "month": 30 * 86_400.0,
}


def _as_epoch(value: Any) -> float | None:
    """Accept the several shapes an exit time arrives in without guessing wrongly.

    KRAKEN_RECONCILED_RESULTS stores epoch seconds as a STRING ('1787173950.17846');
    other tables store ISO-8601. Anything unparseable returns None and the trade is
    reported as `unknown` rather than being dated to now (which would silently pull old
    trades into today's count).
    """
    if value is None:
        return None
    if isinstance(value, (int, float)):
        return float(value)
    text = str(value).strip()
    if not text:
        return None
    try:
        return float(text)
    except ValueError:
        pass
    try:
        cleaned = text.replace("Z", "+00:00")
        parsed = datetime.fromisoformat(cleaned)
        if parsed.tzinfo is None:
            parsed = parsed.replace(tzinfo=timezone.utc)
        return parsed.timestamp()
    except ValueError:
        return None


def _empty_bucket() -> dict[str, Any]:
    return {"successful": 0, "unsuccessful": 0, "breakeven": 0, "unknown": 0, "net_pnl": 0.0}


def summarize_trade_outcomes(trades: Iterable[dict[str, Any]], *, now_epoch: float) -> dict[str, Any]:
    """Bucket closed trades into day/week/month win-loss counts.

    Windows are rolling (last 24h / 7d / 30d) rather than calendar-aligned, so the card
    never shows a near-empty "month" simply because the calendar month just turned over.
    A trade inside the day window is also inside the week and month windows.
    """
    buckets = {name: _empty_bucket() for name in _PERIODS}
    for trade in trades or []:
        if not isinstance(trade, dict):
            continue
        exit_epoch = _as_epoch(trade.get("exit_time") or trade.get("closed_at") or trade.get("updated_at"))
        pnl = trade.get("net_pnl")
        pnl_value: float | None
        try:
            pnl_value = None if pnl is None else float(pnl)
        except (TypeError, ValueError):
            pnl_value = None
        for name, window in _PERIODS.items():
            if exit_epoch is None or now_epoch - exit_epoch > window or exit_epoch > now_epoch + 60:
                continue
            bucket = buckets[name]
            if pnl_value is None:
                bucket["unknown"] += 1
            elif pnl_value > 0:
                bucket["successful"] += 1
                bucket["net_pnl"] += pnl_value
            elif pnl_value < 0:
                bucket["unsuccessful"] += 1
                bucket["net_pnl"] += pnl_value
            else:
                bucket["breakeven"] += 1
    for bucket in buckets.values():
        bucket["net_pnl"] = round(bucket["net_pnl"], 8)
        bucket["settled"] = bucket["successful"] + bucket["unsuccessful"] + bucket["breakeven"]
        bucket["total"] = bucket["settled"] + bucket["unknown"]
        bucket["win_rate"] = (
            round(bucket["successful"] / bucket["settled"], 4) if bucket["settled"] else None
        )
    return buckets


def deterministic_lessons_line(buckets: dict[str, Any]) -> str:
    """An honest one-liner used when no AI summary is available.

    Never invents a lesson. With no settled trades it says exactly that -- which is the
    truthful state for an account that has been correctly declining to trade.
    """
    month = buckets.get("month") or _empty_bucket()
    settled = month.get("settled") or 0
    if not settled:
        unknown = month.get("unknown") or 0
        if unknown:
            return (
                f"No lessons yet: {unknown} trade(s) closed in the last 30 days but none have a "
                "reconciled profit or loss recorded, so none can be judged."
            )
        return "No trades have closed in the last 30 days, so there is nothing to learn from yet."
    wins = month.get("successful") or 0
    losses = month.get("unsuccessful") or 0
    net = month.get("net_pnl") or 0.0
    direction = "ahead" if net > 0 else "behind" if net < 0 else "flat"
    return (
        f"Over the last 30 days {wins} trade(s) made money and {losses} lost money, "
        f"leaving the account {direction} by {abs(net):.2f} overall."
    )


def load_closed_trades(db_path: Path, *, limit: int = 400) -> list[dict[str, Any]]:
    """Closed trades with a real exit time, newest first.

    KRAKEN_RECONCILED_RESULTS is the source of truth for crypto: it is what the AI capital
    ledger itself reconciles against, and it carries a genuine net_pnl including exchange
    fees. Rows still awaiting fill are excluded -- an unfilled exit is not a closed trade.
    Any read failure returns an empty list rather than raising: this powers a Founder
    display card, and a reporting query must never be able to break the briefing.
    """
    try:
        with closing(connect(db_path)) as conn:
            conn.row_factory = sqlite3.Row
            rows = conn.execute(
                """
                SELECT symbol, side, status, exit_time, net_pnl, gross_pnl, net_r
                FROM KRAKEN_RECONCILED_RESULTS
                WHERE exit_time IS NOT NULL AND status = 'closed'
                ORDER BY updated_at DESC
                LIMIT ?
                """,
                (int(limit),),
            ).fetchall()
        return [dict(row) for row in rows]
    except Exception:
        return []


def trade_scorecard(db_path: Path, *, now_epoch: float | None = None) -> dict[str, Any]:
    """The Founder-facing scorecard payload: day/week/month counts plus a lessons line."""
    now = datetime.now(timezone.utc).timestamp() if now_epoch is None else float(now_epoch)
    trades = load_closed_trades(db_path)
    buckets = summarize_trade_outcomes(trades, now_epoch=now)
    return {
        "generated_at": datetime.fromtimestamp(now, tz=timezone.utc).isoformat(),
        "day": buckets["day"],
        "week": buckets["week"],
        "month": buckets["month"],
        "lessons": deterministic_lessons_line(buckets),
        "lessons_source": "counts",
        "closed_trades_considered": len(trades),
    }
